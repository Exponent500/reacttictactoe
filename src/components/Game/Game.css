/* .board-row:after {
    clear: both;
    content: "";
    display: table;
} */

.game {
    width: 90vmin;
    height: 90vmin;
    margin: auto;
    position: absolute;
    right: 0;
    left: 0;
}