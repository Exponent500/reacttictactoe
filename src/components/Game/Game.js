import React, { Component } from 'react';
import Board from '../Board/Board';
import './Game.css';

class Game extends Component {
    constructor(props) {
        super(props);
        this.state = {
            regularMode: true,
            turn: 0,
            player: 1,
            history: [{
                squares: Array(9).fill(null)
            }]
        };
    }

    /**
     * Click handler for when user clicks on one of the squares within the tic tac toe board
     * @param squareIndex - index that represents which square was clicked
     */
    onSquareClick(squareIndex) {
        // make copy of the history
        const history = this.state.history.slice(0, this.state.turn + 1);
        // get latest element in the overall history
        const current = history[history.length - 1];
        const squares = current.squares.slice();
        const currentPlayer = this.state.player;
        const regularMode = this.state.regularMode;

        // guards against click events when:
        // 1. the player clicks on a square that's already been clicked
        // 2. the player clicks on any square after the game has been won
        if (this.hasAPlayerWon(squares) || squares[squareIndex]) {
            return;
        }

        // handles which characters to show for each player based on the mode
        if (currentPlayer === 1 && regularMode) {
            squares[squareIndex] = 'X';
        } else if (currentPlayer === 1 && !regularMode) {
            squares[squareIndex]= 'Y';
        } else if (currentPlayer === 2 && regularMode) {
            squares[squareIndex] = 'O';
        } else if (currentPlayer === 2 && !regularMode) {
            squares[squareIndex] = 'Z'
        }

        // determine the next player
        const nextPlayer = this.state.player === 1 ? 2 : 1;

        this.setState({
            turn: history.length,
            player: nextPlayer,
            history: history.concat({
                squares: squares
            })
        });
    }

    /**
     * Checks to see if a player has won
     * @param squares - array of squares representing the current state of the board
     */
    hasAPlayerWon(squares) {
        // houses all the potential winning lines
        const winningLines = [
            [0, 1, 2],
            [3, 4, 5],
            [6, 7, 8],
            [0, 3, 6],
            [1, 4, 7],
            [2, 5, 8],
            [0, 4, 8],
            [2, 4, 6],
        ];

        // if winning line has been hit, return the winner's value
        for(let i = 0; i < winningLines.length; i++) {
            const [a, b, c] = winningLines[i];
            if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
                return squares[a];
            }
        };
        // if no winning line is hit
        return null;
    }

    /**
     * Click handler for when user wants to undo the last move
     */
    onUndoLastMove() {
        if (this.state.history.length === 1) return;
        // make copy of the history, without the last turn
        const history = this.state.history.slice(0, this.state.turn);
        const previousTurn = history.length - 1;
        const previousPlayer = this.state.player === 1 ? 2 : 1;

        this.setState({
            history: history,
            turn: previousTurn,
            player: previousPlayer
        });
    }

    /**
     * Click handler for when user changes the characters used for the game
     */
    onToggleCharacters() {
        const regularMode = !this.state.regularMode;
        // make copy of the history
        const history = this.state.history.slice(0, this.state.turn + 1);
        // update history of moves so that the appropriate characters are shown based on the mode
        const updatedHistory = history.map( item => {
            const squares = item.squares.map( square => {
                if (square === 'X') return 'Y';
                if (square === 'Y') return 'X';
                if (square === 'O') return 'Z';
                if (square === 'Z') return 'O';
            });
            return { squares: squares };
        });
        this.setState({
            regularMode: regularMode,
            history: updatedHistory
        })
    }

    render() {
        const history = this.state.history;
        const turn = this.state.turn;
        const current = history[turn];
        const currentPlayer = this.state.player;
        const winner = this.hasAPlayerWon(current.squares);

        let status;

        // determines what status is displayed to the players
        if (winner && currentPlayer === 1) {
            status = 'Player 2 wins!';
        } else if (winner && currentPlayer === 2) {
            status = 'Player 1 wins!';
        } else if (!winner && turn === 9) {
            status = 'Tie game!';
        } else if (!winner && turn < 9) {
            status = `Player ${this.state.player}'s turn.`;
        }

        return(
            <div className="game">
                <div className="game-status">
                    {status}
                </div>
                <div className="">
                    <Board
                        squares={this.state.history[this.state.turn].squares}
                        onClick={squareIndex => this.onSquareClick(squareIndex)}
                    ></Board>
                </div>
                <div className="">
                    <button className="btn-undo-move"
                            onClick={() => this.onUndoLastMove()}
                    >Undo Last Move</button>
                </div>
                <div className="">
                    <button className="btn-toggle-characters"
                            type="toggle" 
                            onClick={() => this.onToggleCharacters()}
                    >Toggle Characters Used</button>
                </div>
            </div>
        );
    }
}

export default Game;