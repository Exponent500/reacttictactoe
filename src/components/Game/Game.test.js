import React from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import Game from './Game';

describe('Game', () => {
    it('should render correctly', () => {
        const wrapped = shallow(<Game />); 
        expect(shallowToJson(wrapped)).toMatchSnapshot();
    });

    it('should render game status correctly if it is a brand new game ', () => {
        const wrapped = mount(<Game />);
        const gameStatus = wrapped.find('.game-status').text();
        expect(gameStatus).toEqual("Player 1's turn.");
    });

    it('should render game status correctly if it is player 2 move', () => {
        const wrapped = mount(<Game />);
        let gameStatus;

        // Player 1's turn
        const squareZero = wrapped.find('.square').first();
        squareZero.simulate('click');

        gameStatus = wrapped.find('.game-status').text();
        expect(gameStatus).toEqual("Player 2's turn.");
    });

    it('should render game status correctly if a player won', () => {
        const wrapped = mount(<Game />);
        let gameStatus;

        // Player 1's turn
        const squareZero = wrapped.find('.square').at(0);
        squareZero.simulate('click');

        // Player 2's turn
        const squareOne = wrapped.find('.square').at(1);
        squareOne.simulate('click');
        
        // Player 1's turn
        const squareFour = wrapped.find('.square').at(4);
        squareFour.simulate('click');
        
        // Player 2's turn
        const squareFive = wrapped.find('.square').at(5);
        squareFive.simulate('click');

        // Player 1's turn
        const squareEight = wrapped.find('.square').at(8);
        squareEight.simulate('click');

        gameStatus = wrapped.find('.game-status').text();
        expect(gameStatus).toEqual("Player 1 wins!");
    });

    it('should render game status correctly if it is a tie', () => {
        const wrapped = mount(<Game />);
        let gameStatus;

        // Player 1's turn
        const squareFour = wrapped.find('.square').at(4);
        squareFour.simulate('click');

        // Player 2's turn
        const squareOne = wrapped.find('.square').at(1);
        squareOne.simulate('click');
        
        // Player 1's turn
        const squareTwo = wrapped.find('.square').at(2);
        squareTwo.simulate('click');
        
        // Player 2's turn
        const squareSix = wrapped.find('.square').at(6);
        squareSix.simulate('click');

        // Player 1's turn
        const squareThree = wrapped.find('.square').at(3);
        squareThree.simulate('click');

        // Player 2's turn
        const squareFive = wrapped.find('.square').at(5);
        squareFive.simulate('click');

        // Player 1's turn
        const squareZero = wrapped.find('.square').at(0);
        squareZero.simulate('click');

        // Player 2's turn
        const squareEight = wrapped.find('.square').at(8);
        squareEight.simulate('click');

        // Player 1's turn
        const squareSeven = wrapped.find('.square').at(7);
        squareSeven.simulate('click');

        gameStatus = wrapped.find('.game-status').text();
        expect(gameStatus).toEqual("Tie game!");
    });

    it('should update the Board and Game Status if undo last move button is clicked after moves have been made', () => {
        const wrapped = mount(<Game />);
        const squareFourValue = wrapped.find('.square-value').at(4);
        let gameStatus;

        // Player 1's turn
        const squareFour = wrapped.find('.square').at(4);
        squareFour.simulate('click');

        expect(squareFourValue.text()).toEqual('X');

        wrapped.find('.btn-undo-move').simulate('click');

        expect(squareFourValue.text()).toEqual('');

        gameStatus = wrapped.find('.game-status').text();
        expect(gameStatus).toEqual("Player 1's turn.");

    });

    it('should update the Board if the toggle characters button is clicked after moves have been made', () => {
        const wrapped = mount(<Game />);
        const squareZeroValue = wrapped.find('.square-value').at(0);
        const squareFourValue = wrapped.find('.square-value').at(4);

        // Player 1's turn
        const squareFour = wrapped.find('.square').at(4);
        squareFour.simulate('click');

        expect(squareFourValue.text()).toEqual('X');

        // Player 2's turn
        const squareZero = wrapped.find('.square').at(0);
        squareZero.simulate('click');

        expect(squareZeroValue.text()).toEqual('O');        

        wrapped.find('.btn-toggle-characters').simulate('click');

        expect(squareFourValue.text()).toEqual('Y');
        expect(squareZeroValue.text()).toEqual('Z');

    });

    it('should display alternate characters within squares when new game is started after game mode is switched to use said alternate characters', () => {
        const wrapped = mount(<Game />);
        const squareZeroValue = wrapped.find('.square-value').at(0);

        wrapped.find('.btn-toggle-characters').simulate('click');

        // Player 1's turn
        const squareZero = wrapped.find('.square').at(0);
        squareZero.simulate('click');

        expect(squareZeroValue.text()).toEqual('Y');
    });
});