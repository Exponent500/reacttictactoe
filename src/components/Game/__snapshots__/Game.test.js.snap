// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should render correctly 1`] = `
<div
  className="game"
>
  <div
    className="game-status"
  >
    Player 1's turn.
  </div>
  <div
    className=""
  >
    <Board
      onClick={[Function]}
      squares={
        Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ]
      }
    />
  </div>
  <div
    className=""
  >
    <button
      className="btn-undo-move"
      onClick={[Function]}
    >
      Undo Last Move
    </button>
  </div>
  <div
    className=""
  >
    <button
      className="btn-toggle-characters"
      onClick={[Function]}
      type="toggle"
    >
      Toggle Characters Used
    </button>
  </div>
</div>
`;
